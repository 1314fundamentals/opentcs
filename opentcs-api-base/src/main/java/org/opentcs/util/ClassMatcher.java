// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

import java.io.Serializable;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Predicate;

/**
 * Checks whether the input argument is assignable to any of a given set of classes.
 */
public class ClassMatcher
    implements
      Predicate<Object>,
      Serializable {

  /**
   * The set of classes to evaluate incoming events to.
   */
  private final Set<Class<?>> clazzes = new HashSet<>();

  /**
   * Creates a new instance.
   *
   * @param events The set of classes to evaluate incoming events to.
   */
  public ClassMatcher(Class<?>... events) {
    clazzes.addAll(Arrays.asList(events));
  }

  @Override
  public boolean test(Object object) {
    return clazzes.stream().anyMatch(clazz -> clazz.isAssignableFrom(object.getClass()));
  }

  @Override
  public Predicate<Object> negate() {
    return new ClassMatcher() {
      @Override
      public boolean test(Object t) {
        return !ClassMatcher.this.test(t);
      }
    };
  }
}
