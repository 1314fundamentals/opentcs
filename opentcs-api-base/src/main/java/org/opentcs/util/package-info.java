// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
/**
 * A collection of classes providing miscellaneous functions that are used
 * throughout openTCS.
 */
package org.opentcs.util;
