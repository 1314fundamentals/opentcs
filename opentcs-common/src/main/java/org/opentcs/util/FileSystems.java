// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

import static java.util.Objects.requireNonNull;

import java.io.File;

/**
 * This class provides helper methods for working with file systems.
 */
public final class FileSystems {

  /**
   * Prevents creation of instances.
   */
  private FileSystems() {
  }

  /**
   * Recursively deletes a given file/directory.
   *
   * @param target The file/directory to be deleted recursively.
   * @return <code>true</code> if deleting the target was successful, else
   * <code>false</code>.
   */
  public static boolean deleteRecursively(File target) {
    requireNonNull(target, "target");

    // If the target is a directory, remove its contents first.
    if (target.isDirectory()) {
      File[] entries = target.listFiles();
      for (File curEntry : entries) {
        boolean successful;

        if (curEntry.isDirectory()) {
          successful = deleteRecursively(curEntry);
        }
        else {
          successful = curEntry.delete();
        }

        if (!successful) {
          return false;
        }
      }
    }

    return target.delete();
  }
}
