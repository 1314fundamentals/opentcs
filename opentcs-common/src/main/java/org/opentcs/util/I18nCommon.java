// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

/**
 * Defines constants regarding internationalization.
 */
public interface I18nCommon {

  /**
   * The path to the project's resource bundle.
   */
  String BUNDLE_PATH = "i18n/org/opentcs/common/Bundle";
}
