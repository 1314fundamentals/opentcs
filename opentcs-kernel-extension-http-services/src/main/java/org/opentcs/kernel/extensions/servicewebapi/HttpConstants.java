// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

/**
 * Defines some HTTP-related constants.
 */
public class HttpConstants {

  /**
   * Name of the header that is expected to contain the API access keys.
   */
  public static final String HEADER_NAME_ACCESS_KEY = "X-Api-Access-Key";
  /**
   * Content type for plain text.
   */
  public static final String CONTENT_TYPE_TEXT_PLAIN_UTF8 = "text/plain; charset=utf-8";
  /**
   * Content type for JSON structures.
   */
  public static final String CONTENT_TYPE_APPLICATION_JSON_UTF8 = "application/json; charset=utf-8";

  /**
   * Prevents instantiation.
   */
  private HttpConstants() {
  }
}
