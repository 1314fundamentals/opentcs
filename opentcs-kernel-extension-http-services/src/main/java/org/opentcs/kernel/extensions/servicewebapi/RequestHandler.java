// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import io.javalin.apibuilder.EndpointGroup;
import org.opentcs.components.Lifecycle;

/**
 * A request handler.
 */
public interface RequestHandler
    extends
      Lifecycle {

  /**
   * Creates the handler's routes for the web service.
   */
  EndpointGroup createRoutes();
}
