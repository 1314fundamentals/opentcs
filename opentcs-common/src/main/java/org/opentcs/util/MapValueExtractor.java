// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

import static java.util.Objects.requireNonNull;

import jakarta.annotation.Nonnull;
import java.util.Map;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Provides utility methods for extracting values from {@link Map}s.
 */
public class MapValueExtractor {

  private static final Logger LOG = LoggerFactory.getLogger(MapValueExtractor.class);

  /**
   * Creates a new instance.
   */
  public MapValueExtractor() {
  }

  /**
   * Extracts a string value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @return An {@link Optional} containing the extracted value as a string, or an empty
   * {@link Optional} if the map does not contain an entry with the specified key.
   */
  public Optional<String> extractString(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");

    return Optional.ofNullable(map.get(key));
  }

  /**
   * Extracts an integer value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @return An {@link Optional} containing the extracted value parsed as an integer, or an empty
   * {@link Optional} if the map does not contain an entry with the specified key, or it's value
   * cannot be parsed.
   */
  public Optional<Integer> extractInteger(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");

    return Optional.ofNullable(map.get(key))
        .map(value -> {
          try {
            return Integer.valueOf(value);
          }
          catch (NumberFormatException e) {
            LOG.warn("Could not parse integer value for key '{}': {}", key, value);
            return null;
          }
        });
  }

  /**
   * Extracts a long value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @return An {@link Optional} containing the extracted value parsed as a long, or an empty
   * {@link Optional} if the map does not contain an entry with the specified key, or it's value
   * cannot be parsed.
   */
  public Optional<Long> extractLong(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");

    return Optional.ofNullable(map.get(key))
        .map(value -> {
          try {
            return Long.valueOf(value);
          }
          catch (NumberFormatException e) {
            LOG.warn("Could not parse long value for key '{}': {}", key, value);
            return null;
          }
        });
  }

  /**
   * Extracts a double value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @return An {@link Optional} containing the extracted value parsed as a double, or an empty
   * {@link Optional} if the map does not contain an entry with the specified key, or it's value
   * cannot be parsed.
   */
  public Optional<Double> extractDouble(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");

    return Optional.ofNullable(map.get(key))
        .map(value -> {
          try {
            return Double.valueOf(value);
          }
          catch (NumberFormatException e) {
            LOG.warn("Could not parse double value for key '{}': {}", key, value);
            return null;
          }
        });
  }

  /**
   * Extracts a boolean value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @return An {@link Optional} containing the extracted value parsed as a boolean, or an empty
   * {@link Optional} if the map does not contain an entry with the specified key, or it's value
   * cannot be parsed.
   */
  public Optional<Boolean> extractBoolean(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");

    return Optional.ofNullable(map.get(key))
        .map(Boolean::valueOf);
  }

  /**
   * Extracts an enum value from the given map.
   *
   * @param key The key of the map entry to extract the value from.
   * @param map The map to extract the value from.
   * @param enumClass The class of the enum to extract.
   * @return An {@link Optional} containing the extracted value mapped to an enum of the
   * specified class, or an empty {@link Optional} if the map does not contain an entry with the
   * specified key, or it's value cannot be mapped.
   * @param <E> The type of the enum.
   */
  public <E extends Enum<E>> Optional<E> extractEnum(
      @Nonnull
      String key,
      @Nonnull
      Map<String, String> map,
      @Nonnull
      Class<E> enumClass
  ) {
    requireNonNull(key, "key");
    requireNonNull(map, "map");
    requireNonNull(enumClass, "enumClass");

    return Optional.ofNullable(map.get(key))
        .map(value -> {
          try {
            return Enum.valueOf(enumClass, value);
          }
          catch (IllegalArgumentException e) {
            LOG.warn("Could not map enum value for key '{}': {}", key, value);
            return null;
          }
        });
  }
}
