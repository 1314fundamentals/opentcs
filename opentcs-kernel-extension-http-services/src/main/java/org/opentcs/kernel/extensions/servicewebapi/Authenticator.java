// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import static java.util.Objects.requireNonNull;

import com.google.common.base.Strings;
import io.javalin.http.Context;
import jakarta.inject.Inject;
import java.util.Objects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Authenticates incoming requests.
 */
public class Authenticator {

  /**
   * This class's logger.
   */
  private static final Logger LOG = LoggerFactory.getLogger(Authenticator.class);
  /**
   * Defines the required access rules.
   */
  private final ServiceWebApiConfiguration configuration;

  /**
   * Creates a new instance.
   *
   * @param configuration Defines the required access rules.
   */
  @Inject
  public Authenticator(ServiceWebApiConfiguration configuration) {
    this.configuration = requireNonNull(configuration, "configuration");
  }

  /**
   * Checks whether authentication is required and the given request is authenticated.
   *
   * @param context The context to be checked.
   * @return <code>true</code> if, and only if, authentication is required and the given request is
   * authenticated.
   */
  public boolean isAuthenticated(Context context) {
    requireNonNull(context, "context");

    String requestAccessKey = context.header(HttpConstants.HEADER_NAME_ACCESS_KEY);
    LOG.debug(
        "Provided access key in header is '{}', required value is '{}'",
        requestAccessKey,
        configuration.accessKey()
    );

    // Any empty access key indicates authentication is not required.
    if (Strings.isNullOrEmpty(configuration.accessKey())) {
      LOG.debug("No access key, authentication not required.");
      return true;
    }

    return Objects.equals(requestAccessKey, configuration.accessKey());
  }

}
