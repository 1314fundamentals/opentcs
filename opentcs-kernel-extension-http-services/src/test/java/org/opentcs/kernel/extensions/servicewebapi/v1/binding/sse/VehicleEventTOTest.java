// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;

/**
 * Tests for {@link VehicleEventTO}.
 */
class VehicleEventTOTest {

  private JsonBinder jsonBinder;

  @BeforeEach
  void setUp() {
    jsonBinder = new JsonBinder();
  }

  @Test
  void jsonSample() {
    VehicleEventTO to = new VehicleEventTO(
        createVehicle(VehicleEventTO.IntegrationLevelTO.TO_BE_UTILIZED),
        createVehicle(VehicleEventTO.IntegrationLevelTO.TO_BE_RESPECTED)
    );

    Approvals.verify(jsonBinder.toJson(to));
  }

  private VehicleEventTO.VehicleTO createVehicle(
      VehicleEventTO.IntegrationLevelTO integrationLevel
  ) {
    Instant time = Instant.parse("2025-07-22T12:03:00Z");
    return new VehicleEventTO.VehicleTO()
        .setName("vehicle-1")
        .setProperties(new TreeMap<>(Map.of("key-1", "value-1", "key-2", "value-2")))
        .setHistory(
            new ObjectHistoryTO().setEntries(
                List.of(
                    new ObjectHistoryTO.ObjectHistoryEntryTO()
                        .setEventCode("event-code-1")
                        .setTimestamp(time)
                        .setSupplements(List.of("supplement-1", "supplement-2"))
                )
            )
        )
        .setBoundingBox(
            new VehicleEventTO.BoundingBoxTO()
                .setLength(1)
                .setWidth(2)
                .setHeight(3)
                .setReferenceOffset(new VehicleEventTO.CoupleTO().setX(4).setY(5))
        )
        .setEnergyLevelThresholdSet(
            new VehicleEventTO.EnergyLevelThresholdSetTO()
                .setEnergyLevelCritical(1)
                .setEnergyLevelGood(2)
                .setEnergyLevelSufficientlyRecharged(3)
                .setEnergyLevelFullyRecharged(4)
        )
        .setMaxVelocity(200)
        .setMaxReverseVelocity(100)
        .setRechargeOperation("recharge-operation")
        .setProcState(
            new VehicleEventTO.TimestampedVehicleProcStateTO()
                .setProcState(VehicleEventTO.ProcStateTO.IDLE)
                .setTimestamp(time)
        )
        .setTransportOrder("transport-order-1")
        .setOrderSequence("order-sequence-1")
        .setAcceptableOrderTypes(
            List.of(
                new VehicleEventTO.AcceptableOrderTypeTO()
                    .setName("acceptable-order-type-1")
                    .setPriority(2),
                new VehicleEventTO.AcceptableOrderTypeTO()
                    .setName("acceptable-order-type-2")
                    .setPriority(5)
            )
        )
        .setClaimedResources(
            List.of(
                List.of("resource-1", "resource-2")
            )
        )
        .setAllocatedResources(
            List.of(
                List.of("resource-3", "resource-4"),
                List.of("resource-5", "resource-6")
            )
        )
        .setState(
            new VehicleEventTO.TimestampedVehicleStateTO()
                .setState(VehicleEventTO.StateTO.IDLE)
                .setTimestamp(time)
        )
        .setIntegrationLevel(integrationLevel)
        .setPaused(true)
        .setCurrentPosition("position-1")
        .setPose(
            new VehicleEventTO.PoseTO()
                .setPosition(
                    new VehicleEventTO.TripleTO()
                        .setX(1)
                        .setY(2)
                        .setZ(3)
                )
                .setOrientationAngle(4.5)
        )
        .setEnergyLevel(42)
        .setLoadHandlingDevices(
            List.of(
                new VehicleEventTO.LoadHandlingDeviceTO()
                    .setLabel("lhd-1")
                    .setFull(true)
            )
        )
        .setEnvelopeKey("envelope-key-1")
        .setLayout(
            new VehicleEventTO.LayoutTO()
                .setRouteColor(
                    new VehicleEventTO.LayoutTO.ColorTO()
                        .setRed(60)
                        .setGreen(20)
                        .setBlue(40)
                )
        );
  }
}
