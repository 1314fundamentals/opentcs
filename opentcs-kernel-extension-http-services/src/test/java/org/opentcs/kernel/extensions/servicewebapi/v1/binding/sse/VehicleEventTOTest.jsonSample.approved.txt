{
  "currentObjectState" : {
    "name" : "vehicle-1",
    "properties" : {
      "key-1" : "value-1",
      "key-2" : "value-2"
    },
    "history" : {
      "entries" : [ {
        "timestamp" : "2025-07-22T12:03:00Z",
        "eventCode" : "event-code-1",
        "supplements" : [ "supplement-1", "supplement-2" ]
      } ]
    },
    "boundingBox" : {
      "length" : 1,
      "width" : 2,
      "height" : 3,
      "referenceOffset" : {
        "x" : 4,
        "y" : 5
      }
    },
    "energyLevelThresholdSet" : {
      "energyLevelCritical" : 1,
      "energyLevelGood" : 2,
      "energyLevelFullyRecharged" : 4,
      "energyLevelSufficientlyRecharged" : 3
    },
    "energyLevel" : 42,
    "maxVelocity" : 200,
    "maxReverseVelocity" : 100,
    "rechargeOperation" : "recharge-operation",
    "loadHandlingDevices" : [ {
      "label" : "lhd-1",
      "full" : true
    } ],
    "state" : {
      "state" : "IDLE",
      "timestamp" : "2025-07-22T12:03:00Z"
    },
    "procState" : {
      "procState" : "IDLE",
      "timestamp" : "2025-07-22T12:03:00Z"
    },
    "integrationLevel" : "TO_BE_UTILIZED",
    "paused" : true,
    "transportOrder" : "transport-order-1",
    "orderSequence" : "order-sequence-1",
    "acceptableOrderTypes" : [ {
      "name" : "acceptable-order-type-1",
      "priority" : 2
    }, {
      "name" : "acceptable-order-type-2",
      "priority" : 5
    } ],
    "claimedResources" : [ [ "resource-1", "resource-2" ] ],
    "allocatedResources" : [ [ "resource-3", "resource-4" ], [ "resource-5", "resource-6" ] ],
    "currentPosition" : "position-1",
    "pose" : {
      "position" : {
        "x" : 1,
        "y" : 2,
        "z" : 3
      },
      "orientationAngle" : 4.5
    },
    "envelopeKey" : "envelope-key-1",
    "layout" : {
      "routeColor" : {
        "red" : 60,
        "green" : 20,
        "blue" : 40
      }
    }
  },
  "previousObjectState" : {
    "name" : "vehicle-1",
    "properties" : {
      "key-1" : "value-1",
      "key-2" : "value-2"
    },
    "history" : {
      "entries" : [ {
        "timestamp" : "2025-07-22T12:03:00Z",
        "eventCode" : "event-code-1",
        "supplements" : [ "supplement-1", "supplement-2" ]
      } ]
    },
    "boundingBox" : {
      "length" : 1,
      "width" : 2,
      "height" : 3,
      "referenceOffset" : {
        "x" : 4,
        "y" : 5
      }
    },
    "energyLevelThresholdSet" : {
      "energyLevelCritical" : 1,
      "energyLevelGood" : 2,
      "energyLevelFullyRecharged" : 4,
      "energyLevelSufficientlyRecharged" : 3
    },
    "energyLevel" : 42,
    "maxVelocity" : 200,
    "maxReverseVelocity" : 100,
    "rechargeOperation" : "recharge-operation",
    "loadHandlingDevices" : [ {
      "label" : "lhd-1",
      "full" : true
    } ],
    "state" : {
      "state" : "IDLE",
      "timestamp" : "2025-07-22T12:03:00Z"
    },
    "procState" : {
      "procState" : "IDLE",
      "timestamp" : "2025-07-22T12:03:00Z"
    },
    "integrationLevel" : "TO_BE_RESPECTED",
    "paused" : true,
    "transportOrder" : "transport-order-1",
    "orderSequence" : "order-sequence-1",
    "acceptableOrderTypes" : [ {
      "name" : "acceptable-order-type-1",
      "priority" : 2
    }, {
      "name" : "acceptable-order-type-2",
      "priority" : 5
    } ],
    "claimedResources" : [ [ "resource-1", "resource-2" ] ],
    "allocatedResources" : [ [ "resource-3", "resource-4" ], [ "resource-5", "resource-6" ] ],
    "currentPosition" : "position-1",
    "pose" : {
      "position" : {
        "x" : 1,
        "y" : 2,
        "z" : 3
      },
      "orientationAngle" : 4.5
    },
    "envelopeKey" : "envelope-key-1",
    "layout" : {
      "routeColor" : {
        "red" : 60,
        "green" : 20,
        "blue" : 40
      }
    }
  }
}